### ISSUE TITLE
Verse Change Notification system

### ISSUE THREAD

**Issue by benjore** [unfoldingWord/translationCore#3562]
## Story

As a user, I want to know if I have inadvertently invalidated a check in any tool without opening the tool so that I can finish all my work without constantly opening and closing tools. 

## Specs 

- [ ] Each tool card has a badge with the number of invalidated checks. This updates every time the user navigates to the tools page. 
- [ ] Definition of "invalidation"
  - [ ] For the tW tool, if something is changed in a current selection. 
  - [ ] For the wA tool, if any word/token is changed, then that word is invalidated (and the link is thrown for the whole verse).

**Comment by jag3773** [comment]
Related to #3559 and #3086

**Comment by klappy** [comment]
Request to persist the invalidated checks since it can take a long time to refresh.
Potentially add a reducer to persist invalidated checks for the project/tool combo.
Clicking refresh updates that number.

**Comment by mannycolon** [comment]
@benjore Whats the definition of an invalidated check for tW and wA?

**Comment by mannycolon** [comment]
@benjore These mockups need a description.
What does the yellow question badge mean? and when is it supposed to show up?

**Comment by benjore** [comment]
@mannycolon 

Definition of invalidated check for tW: if the text of a selection has been edited. Example: if the selection is "the word" and is then edited to be "the Word", that check becomes invalidated.

Definition of invalidated check for wA: if any text in a verse is edited, the alignment for that verse becomes invalidated. 

Yellow badges: when a user goes to the tools cards step of the stepper, the app will automatically calculate the number of verses changed. The app will not automatically calculate the number of invalidated checks (as there couple be way more). The app will show the yellow badge in this case. If the user clicks the circular arrow, the app will then calculate any invalidated checks and then display a red badge.

**Comment by mannycolon** [comment]
# Tasks
- [3] Create functions/methods/actions to retrieve invalidated checks for tW
- [3] Create functions/methods/actions to retrieve invalidated checks for wA
- [2] Create functions/methods/actions to retrieve edited verses

**Comment by mannycolon** [comment]
Per conversation with @benjore 
- No need to show edited verses badge
- No need for refresh button.

**Comment by cckozie** [comment]
0.10.0-7f20ffa
[en_tit 2.zip](https://github.com/unfoldingWord-dev/translationCore/files/1976059/en_tit.2.zip)
Project initially shows no broken links on tW card and 9 broken links on wA card
1. Open project in tW
2. Select 'an apostle' for the apostle check in 1:1
3. Switch to alignment tool
4. Edit 1:1 to change 'an apostle' to 'a apostle' (delete the 'n' in 'an')
5. Click the tool icon to go to the tools page
BUG: The tW card should now show that a check has been invalidated, but it does not show that
**[THIS PROBLEM IS FIXED IN 1.1.3 (4817b07)]**
At this point the tW card shows 1 broken link, and the wA card shows 8 broken links. (Interesting since 1 more alignment was invalidated, but expected as the broken link in wA gets cleared when the verse is active)

(I SUSPECT THE FOLLOWING WILL BE FIXED BY #4433) **It is fixed in 1.1.3 (4817b07)**
6. Reopen the project in tW. An alert properly notifies the user that a selection has been invalidated.
7. Edited the verse again in expanded scripture pane to change the 'a' back to 'an'
(The broken link for 1:1 now shows the number 2)
8. Click on the tool icon to return to the tools page
9. Both tools now show 1 broken link
10. Launch the project in alignment tool
BUG: ERROR is displayed where the scripture pane should be
![image](https://user-images.githubusercontent.com/8171759/39656074-1f8cd07c-4fc3-11e8-98ac-eb07b33668f3.png)

**Comment by mannycolon** [comment]
@cckozie The bug described above is not related to this issue

**Comment by mannycolon** [comment]
Blocked by wA tool refactor #4319

**Comment by mannycolon** [comment]
Initial estimate: 3 hours
Actual: 8 hours

**Comment by mannycolon** [comment]
![bitmoji](https://render.bitstrips.com/v2/cpanel/702e8c08-428f-42eb-9a8a-34779db688c9-3829d733-47ee-457c-a647-dcc18418282b-v1.png?transparent=1&palette=1&width=246)

**Comment by cckozie** [comment]
0.10.0-9d56594 (undefined) 
A. I followed steps 1-5 above on a newly imported untouched USFM project. Upon returning to the tools page, the tW card shows 1 invalidated check as expected. **(This working as expected)**
![image](https://user-images.githubusercontent.com/8171759/40927899-88683c3c-67e5-11e8-9e21-43316ca4c197.png)
When the project is opened in tW, the check card for apostle shows 2 checks invalidated instead of 1.
![image](https://user-images.githubusercontent.com/8171759/40928336-ace6305e-67e6-11e8-8846-84d89a76e944.png)

B. After reselecting to fix the invalidated check, the number badge still appears on the check in the menu. (It does not show on the tW card on the tools page)
**This is no longer a problem in 1.1.3 (4817b07)**
C. Delete the project and reimport from USFM. Then follow these steps:
1. Open the project with tW.
2. Make a selection for the first check and click Save.
3. Expand the scripture pane.
4. Make an edit to a different verse in the chapter and click Save.
5. Click on the Tool icon to go to the Tools page.
Notice that both the wA and tw card indicate invalided alignments/checks even though neither is true. It appears as though the indicator on the cards is counting the number of edits instead of the number of invalidations.
**This is fixed in 1.1.3 (4817b07)**
@benjore When going back into a project in tW after editing in wA has caused invalidated checks, should the "Some selections are no longer valid" alert still be displayed as before and when edited in tW? It seems like the app should be consistent in displaying that alert, or not, regardless of where the editing was performed.

**Comment by benjore** [comment]
@cckozie I'm not totally following your example, but on the whole, it seems like consistency is what we want.

**Comment by cckozie** [comment]
@benjore If I edit a verse in tW and it causes the selection to become invalid, I get the alert that warms me about that. That is the expected (past) behaviort.
However, If I make the edit in wA, I do not get that alert either in wA or when I go back to tW. It does show the invalid icon, but I never get the alert. Option are:
A. Leave it as is
B. Generate the alert in wA
C. Generate the alert in tW
D. Both B. and C.
**D. is the case in 1.1.3 (4817b07) as expected**

**Comment by mannycolon** [comment]
@cckozie 
> When the project is opened in tW, the check card for apostle shows 2 checks invalidated instead of 1.

What do you mean it shows 2  checks invalidated instead of 1 ?

@mannycolon The number "2" shows in the badge by the invalidated icon

**Comment by mannycolon** [comment]
> B. After reselecting to fix the invalidated check, the number badge still appears on the check in the menu. (It does not show on the tW card on the tools page)

Not sure what this means
@mannycolon Go to the verse with the invalidated check and make a selection so that the check is once again valid. The icon and badge still indicate that it is invalidated.

**Comment by mannycolon** [comment]
> The number "2" shows in the badge by the invalidated icon

@cckozie But it is still one invalidated check even if the icon shows a number 
@mannycolon Sorry, my bad. I kept thinking that the 2 was somehow indicating 2 invalidations rather than 2 icons on the check.

**Comment by cckozie** [comment]
On build 0.10.0 (af7b11b), after making an edit in tW on an aligned verse, I do get the alert stating that the alignments have been reset. However, when I go back to the Tools page, the wA card does not indicate a reset alignment. (The tW card does indicate an invalidated check)
![image](https://user-images.githubusercontent.com/8171759/40994578-37024f76-68c2-11e8-9e46-b3d4abeb561c.png)

**Comment by cckozie** [comment]
Testing on build 0.10.0 (e6fe53e)

1. [x] From the comment on May 4, build 0.10.0-7f20ffa
  There are no remaining issues related to this comment.

2. [x] From the comment 9 days ago, build 0.10.0-9d56594
Issues described in A. and B. are not a problem. C. does remain as a problem. To replicate:
a) Import a new usfm titus project
b) Make a select for a check
c) Open the expanded scripture pane and edit a DIFFERENT verse in the same chapter as the check that was made
d) BUG: When the edit is saved, the Some selections are no longer valid... alert is displayed
e) BUG: On the tools page, both tW and wA incorrectly display 1 broken alignment/check

3. [ ] From the comment 8 days ago, build 0.10.0 (af7b11b)
There is still a problem here. To recreate:
    - [ ] **Part 1**
    1) Delete and reimport the usfm titus project
    2) Load it in wA and align verse 1:1
    3) Switch over to tW and delete the word 'Jesus' verse 1:1
    4) The reset alignment alert is displayed when the Save button is clicked
    5) Go to the tools page
    6) BUG: No indication of reset alignment on the wA card
    7) Load the project in wA and navigate to verse 1:1
    8) BUG: The reset alignment warning is not displayed
    - [ ] **Part 2**
    9) Align verse 1:10
    10) Switch over to tW and change the word 'people' to 'folks in verse 1:10
    11) BUG: The reset alignment warning is not displayed when the Save button is clicked
    12) Go the tools page and see that the wA card does show 1 reset alignment
    13) Load the project in wA and navigate to verse 1:10
    14) The reset alignment alert is correctly displayed

**Comment by mannycolon** [comment]
@cckozie 
For **Part 1**
I'm not sure if the alignment is exactly invalidated. The word `Jesus` was removed but:
**Does that mean that the alignment is invalidated?**
 - Could be the case and could not be the case it depends on the scenario.

**I know the word is missing but would the aligner have to do anything to that alignment to make it a valid alignment?**
- Could be the case and could not be the case it depends on the scenario.

Thus, I'm not sure how we should handle this scenario since it's not clear if it should even be considered an invalidated alignment.

@mannycolon Yes, those are 'invalidations' as correctly indicated by the alert being displayed when the edit is saved. The alignment is not complete until all Greek words have one or more words associated with them and all of the words in the word list have been aligned. Adding, changing, or deleting words in a verse all 'interfere with' alignments.
In a more general sense, anytime the reset alignment message is displayed after an edit in tW, that information should be displayed on the wA tool card and in the wA tool.

**Comment by mannycolon** [comment]
@cckozie @neutrinog 
I uncovered that `Part 2` should be part of the wA tool rewrite story. The API is not properly validating the verse. 

When I changed the word 'people' to 'folks` in Titus 1:10 I uncover the following in getIsValid() (`reducers/alignments/Verse.js`):
```js

targetText =  "For folks there are many rebellious empty talkers and deceivers especially those of the circumcision" 

targetBaselineText = "For there are many rebellious people empty talkers and deceivers especially those of the circumcision"
```

targetText should be `For there are many rebellious folks empty talkers and deceivers especially those of the circumcision` not `For folks there are many rebellious empty talkers and deceivers especially those of the circumcision`

@mannycolon I'm not sure I follow all of that, but if you believe it belongs on the other story, please copy the appropriate info to it.

**Comment by da1nerd** [comment]
@mannycolon @cckozie the reset alignment warning dialog will only appear if your context is the same as the verse you edited.

e.g. navigate to here either before or after you edit 1:10 and you'll see the dialog.
![image](https://user-images.githubusercontent.com/166412/41394617-77855022-6f5f-11e8-9914-40f722b9a1a8.png)

**Comment by cckozie** [comment]
Discussion among @benjore, @mannycolon, @klappy, @cckozie (Guys, feel free to edit as needed)
This writeup pertains to Part 1 in the comments above.

We decided that there currently really isn't a way to clearly define when a verse's alignment should be 'invalidated'. As a group we decided that a better approach would be to add UI element(s) to allow the user to indicate that the alignment is 'complete'. tC will timestamp and store that decision. If edits are later made to the verse, tC will then consider the alignment to no longer be 'complete' or 'valid'. The user will receive that message at the time of editing, and it will be counted as 'broken' on the wA tool card. When the user again visits the verse, many alignment changed (if necessary), and then again indicates that the alignment is 'complete', the counter of 'broken' alignments on the tool card will be decremented.

A new issue will be created to provide this changed functionality. For the purposes of this issue, Part 1 above will no longer be considered a defect.

**Comment by cckozie** [comment]
0.10.0 (ed1e141)
The number of broken checks shown on the tW card is not always accurate.
Scenario 1:
a. Import a Titus USFM project and open in tW
b. Select 'an apostle' in 1:1 and Save
c. Navigate to the 'chosen one...' group
d. Select 'of chosen people' in 1:1 and Save
e. Open 1:1 for editing
f. Change 'an' to 'a'
g. Change 'chosen people' to 'called people'
h. Click Save
g. Go to the Tools page. The tW card shows 1 invalidated check instead of 2.

Scenario 2:
a. Repeat steps a-d above
b. Switch to the word alignment tool
c. Edit 1:1 to delete the word 'an' to invalidate the check and Save
d. Go to the Tools page. The tW card shows 0 invalidated checks.

**Comment by da1nerd** [comment]
Hooked up the progress and invalidation to the tW api.

![peek 2019-01-09 21-09](https://user-images.githubusercontent.com/166412/50947387-ed24da80-1452-11e9-80d7-13ac11f01099.gif)

**Comment by PhotoNomad0** [comment]
Verified in https://tc-builds.door43.org/build/develop/translationCore-macos-x64-1.1.0-7c2a84b.dmg

**Comment by cckozie** [comment]
1.1.0 (7c2a84b)
I suspect that it's more of a problem of this not being finished (at least in the case of tW) rather than it has bugs, but since it was in QA I am failing it.
## Specs 

- [ ] Each tool card has a badge with the number of invalidated checks. This updates every time the user navigates to the tools page. 
- [ ] Definition of "invalidation"
  - [ ] For the tW tool, if something is changed in a current selection. 
  - [ ] For the wA tool, if any word/token is changed, then that word is invalidated (and the link is thrown for the whole verse).

**Comment by cckozie** [comment]
THREE NEW ISSUES WERE CREATED TO HANDLE THESE 3 PROBLEMS SO THAT THE ISSUE COULD BE APPROVED FOR 1.1 (#5685, #5686, #5687)
1.1.0 (1034612)
Scenario A: (still testing other scenarios)
App indicates reset alignments when edits are done that do not actually affect alignments
1. Import https://git.door43.org/qa99/en_full_tit_book
2. Load the project in wA
3. Unalign some of the words as shown here:
![image](https://user-images.githubusercontent.com/8171759/51210178-bbd75f00-18d7-11e9-9251-05be8cc4e723.png)
4. Load the project in tW
5. In verse 1:1 under 'apostle', change the text "an apostle" to "the apostle", give it a reason, then Save
6. Note that the 'some selections no longer valid' alert is displayed
7. Note that when that alert is dismissed, the invalidated alignments alert is incorrectly displayed.
8. Return to the tools page and note that the wA tool card incorrectly indicates that there is one broken alignment.

Scenario B:
When a project is edited external to tC and then reimported, broken alignments are handled properly (the alignments invalidated alert it issues during the import and the broken link number is correct on the tool card). Edits that affect tW selections are not caught until the user navigates to the check that has been invalidated. (The alert is not displayed and the broken link number is not updated during the reimport)
To test:
Use a project that has some alignments and some completed checks.
Export the project to USFM not including the alignment data
Make edits to the USFM file that will break alignments, break selections, and break both.
Import the USFM file giving it the same translation ID as the original, and choose the Overwrite options

Scenario C:
When edits are made in wA that affect tW selections, the selections invalidated alert is not displayed until the user opens the project in tW. The broken link number on the tW card is set correctly.

**Comment by da1nerd** [comment]
I've confirmed that wA is inaccurately indicating alignments have changed due to the propagating effect of word occurrences.
e.g. Since there were other `the`s aligned in the verse their occurrence number changed which invalidated the alignment.

Resolving this is tricky because we need to make sure that rearranging words still causes an invalidation. Instead of doing a straight diff on the two alignment versions I think we can solve this by
1. performing a textual comparison of the alignments
2. performing a relative position comparison of the alignments.

**Comment by PhotoNomad0** [comment]
Verified wrong validation fix (https://github.com/translationCoreApps/wordAlignment/pull/148) in https://tc-builds.door43.org/build/release-v1.1.1/translationCore-macos-x64-1.1.1-000d36b.dmg

**Comment by cckozie** [comment]
Holding until the child issues are through QA.

**Comment by cckozie** [comment]
Problems found in 1.1.3 (e1fae24) 
[45-ACT.usfm.txt.zip](https://github.com/unfoldingWord-dev/translationCore/files/2953907/45-ACT.usfm.txt.zip)
Import this usfm3 project and load it in tW

ISSUE 1, Process A: [FIXED IN 1.1.3 (9d5ff9e)]
Navigate to the 'die' group
Click on verse 7:15
Click the Cancel button
Click the Edit Verse button
Insert the word 'there' between 'die' and its following comma
Choose a reason code
Click the Save button
Notice that the alignments have been reset alert is not displayed, but it should be. **Fixed in 1.1.3 (4817b07)**
Click on the Tool icon in the menu bar. Notice that there is no indication of reset alignments on the wA card. **Fixed in 1.1.3 (4817b07)**
Load the project in wA
Open chapter 7
Notice that the broken link icon is not displayed on verse 15 **Fixed in 1.1.3 (4817b07)**
(The problem of the edit icon not being displayed there is written up in #5832)

ISSUE 1, Process B: (This operates correctly)
Reload the project in tW
Click on verse 9:37
Select the word 'died'
Click the Save button
Click the Edit Verse button
Change the word 'died' to 'expired'
Choose a reason code
Click the Save button
Notice that the some selections are no longer valid alert is displayed, and when it is dismissed, the alignments reset alert is displayed. This is the correct behavior.
Click on the tool icon in the menu bar
Notice that now the wA card indicates 1 broken alignment [As of 1.1.3 (9d5ff9e)] it indicates 2 broken alignments **Works correctly in 1.1.3 (4817b07)**
______________________________________
ISSUE 2: (Similar to issue 1 above, but using the expanded scripture pane for editing)
 [FIXED IN 1.1.3 (9d5ff9e)]
Select verse 13:28
Highlight the words 'for death'
Click the Save button
Click the Save & Continue button
Open the expanded scripture pane
Open verse 13:28 for editing
Insert the word 'his' between 'for' and 'death'
Click the Next button
Choose a reason
Click the Save button
Notice that the alignments reset alert is not issued.
**Fixed in 1.1.3 (4817b07)**
________________________________
ISSUE 3A:  [NOT FIXED IN 1.1.3 (9d5ff9e)]
It appears that there are two different selections reset alerts that are issued sequentially. The first one does not have the tC logo and after about a second it is overlaid by one that does.
[dup reset selections.mov.zip](https://github.com/unfoldingWord-dev/translationCore/files/2954008/dup.reset.selections.mov.zip)
ISSUE 3B:
The same appears to be true for the alignments reset alert when edits are performed in wA.

**Note:** to replicate - this is an edit of acts 4:2 under die - started with the selection 'the dead`, and verse edit to change to 'those dead'
**Fixed in 1.1.3 (4817b07)**
____________________________
ISSUE 4: [FIXED IN 1.1.3 (9d5ff9e)] - but there is now a long delay between the selections reset and alignments invalidated messages with no indication that anything is happening. The delay is ~10 seconds in 1.1.3, whereas it was virtually instantaneous in 1.1.2. I've also found that frequently the OK button on the invalidated alignment alert is unresponsive until about 2 seconds after it is displayed.]
[delay113.mov.zip](https://github.com/unfoldingWord-dev/translationCore/files/2983875/delay113.mov.zip)
[delay112.mov.zip](https://github.com/unfoldingWord-dev/translationCore/files/2983878/delay112.mov.zip)

Import the Acts project again giving it a different translation id
Load the project in tW
Navigate to the 'die' group
Select verse 2:24 if it is not already selected
Highlight the words 'of death'
Click the Save button 
Click the Save & Continue button
Click on the tool icon in the menu bar
Load the project in wA
Navigate to chapter 2
Open the expanded scripture pane
Open verse 2:24 for editing
Change the word 'death' to 'dieing'
Click the Next button
Choose a reason
Click the Save button
Notice that the selections reset alert is not issued
Click on the tool icon in the menu bar
Notice that the tW card does correctly indicate one reset selection
**Fixed in 1.1.3 (4817b07) by addition of checking for invalidations alert**
______________________
ISSUE 5:
I have not been able to consistently replicate this one, but occasionally when I perform an edit that both invalidates selections and alignments, the selections reset alert is automatically dismissed without clicking the OK button, and then the reset alignments one is displayed. 
**I have not seen this behavior in 1.1.3 (4817b07)**

**Comment by PhotoNomad0** [comment]
**Note:**

- Looking at Issue 1 & 2 - it looks like adding a word is not causing the alignments reset alert.
- On issue 3 ~it looks like the dialog is not completely rendered initially until the verse edit notifications are completed.~  Looks like the problem is that WA is issuing two alignments invalidated popups.  The first occurs before the selections validation is complete.  Which causes the flashing.

**Comment by da1nerd** [comment]
@cckozie with regards to Issue 1 & 2: can you confirm that inserting a word should cause the alignments to become invalidated? By design inserting words never cause an invalidation. Could you check with the stakeholders to see if that specification has changed?

**Comment by PhotoNomad0** [comment]
@benjore @cckozie - In looking into https://github.com/unfoldingword-dev/translationcore/issues/3562, issue 1A, it looks like WA has never considered adding a new word to a verse to be invalidation of alignments.  We can add it, but it will need some refactoring of WA to add additional checking based on word bank and complete flag.  Also it will increase the checking time when we open a project and when we launch the WA tool and all verses are checked - guessing between 20 to 50% time increase, but won’t know for sure until we implement.  How do you want to proceed?

**Comment by PhotoNomad0** [comment]
Verfied in https://tc-builds.door43.org/build/release-v1.1.3/translationCore-macos-x64-1.1.3-9d5ff9e.dmg

Issues 1 to 5 should be working.

**Comment by cckozie** [comment]
1.1.3 (9d5ff9e) 
Issues 1, 2, and 4 are fixed. I'm still testing 5.
Issue 3 has not been fixed.
[no_logo.mov.zip](https://github.com/unfoldingWord-dev/translationCore/files/2980930/no_logo.mov.zip)

**Comment by PhotoNomad0** [comment]
@cckozie - yes I, am seeing this also sometimes.  ~Looking like a react artifact related to the animation beginning before the image is drawn.~  Looking like a material UI rendering problem when word-map is running in the background.  Investigating this and asking for help from others.

**Comment by PhotoNomad0** [comment]
**Note:**  Investigating options to resolve issue 3 (sometimes fails):
- disable the animation (dialog drop down) of the alert - so far all attempts to do so have failed.
- tried adding delays around the display of the dialog, but it didn't help.
- tried direct loading of image data (base64) - didn't fix problem.
- tried updating material-UI version from 0.20.0 to 0.20.2 - didn't fix problem.
- set width of image - don't see the dialog resizing after drop down, but image still loads later.
- trying switch to divs.

**Other options to try:**
- can update material-UI version for the alert.
- pause word-map until dialog is completely rendered.

**Comment by PhotoNomad0** [comment]
Tested issue 3 fix with https://tc-builds.door43.org/build/release-v1.1.3/translationCore-macos-x64-1.1.3-2c25b41.dmg

**Comment by cckozie** [comment]
Issue 3 is fixed in 1.1.3 (2c25b41).
Regarding issue 4, now the selections reset alert does disappear within a reasonable time, but there is a 6-8 second delay before the invalidated alignments alert and/or the edit icon on the verse is displayed. During this delay the app is unresponsive to any user input but there is no visual indication to alert the user of that.

**Comment by cckozie** [comment]
1.1.3 (ef323b2)
The delay between when the selections reset alert disappears and the invalidated alignments and/or edit icon are displayed is not about 1/2 of what it was, generally around 4 seconds in Acts.

**Comment by cckozie** [comment]
1.1.3 (ee4a767) 
The delay is down below 4 seconds now, even in the worst case in Acts where there are 14 checks related to 3:13.
('abraham', 'betray', 'father', 'glory', 'god', 'god', 'god', 'godthefather', 'isaac', 'jacob', 'jesus', 'judge', 'pilate', 'servant')

**Comment by PhotoNomad0** [comment]
Verified in https://tc-builds.door43.org/build/release-v1.1.3/translationCore-win-x64-1.1.3-b525e6c.setup.exe

**Comment by cckozie** [comment]
1.1.3 (126c679) 
[reset alignments 3.mov.zip](https://github.com/unfoldingWord-dev/translationCore/files/3031007/reset.alignments.3.mov.zip)
@benjore - I would prefer verbiage something like "Testing for invalidated checks and alignments" rather than "Doing Invalidation Checking" on the first alert. I also don't like the the alignments invalidated alert with its buttons being displayed behind the first alert before it is dismissed.
What say you? Is it OK the way it is, or should we make changes now or write them up for the future?

**Comment by cckozie** [comment]
Per Ben - Let's pass this one for now (assuming all the functionality works). I'll write up a new issue with the desired modifications. (#5941)

**Comment by cckozie** [comment]
All issues are either fixed or covered in other issues as of 1.1.3 (4817b07)
